data/src/nonempty/map.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{
+    borrow::Borrow as _,
+    collections::{BTreeMap, HashMap},
+    convert::TryFrom,
+    hash::{BuildHasher, Hash},
+    marker::PhantomData,
+    ops::Deref,
+};
+
+/// Alias for a [`NonEmptyMap`] backed by a [`BTreeMap`]
+pub type NonEmptyOrderedMap<K, V> = NonEmptyMap<K, V, BTreeMap<K, V>>;
+
+/// Alias for a [`NonEmptyMap`] backed by a [`HashMap`]
+pub type NonEmptyHashMap<K, V, S> = NonEmptyMap<K, V, HashMap<K, V, S>>;
+
+/// Simple newtype wrapper around a map type which witnesses that the map
+/// contains at least one element.
+///
+/// Non-mutating methods of the underlying map are available through the
+/// [`Deref`] impl. Additionally, mutating methods which cannot leave an empty
+/// map behind are provided. To construct a [`NonEmptyMap`], use the [`TryFrom`]
+/// impl.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct NonEmptyMap<K, V, Inner = BTreeMap<K, V>> {
+    inner: Inner,
+    _marker: PhantomData<(K, V)>,
```

Comment:
The `PhantomData` is here to make the compiler not complain about unused `K, V`, right?

Reply:
Yeah Rust. What I'd actually wanted was something like this:

```rust
trait IsEmpty {
    fn is_empty(&self) -> bool;
}

pub struct NonEmpty<T>(T);

impl<T: IsEmpty> TryFrom<T> for NonEmpty<T> {
    type Error = TheEmptinessInside;

    fn try_from(t: T) -> Result<Self, Self::Error> {
        if t.is_empty() { Err(TheEmptinessInside) } else { Ok(Self(t)) }
    }
}
```

but that seems to create many-a-headache with coherence.

Reply:
Hah, wait. If we don't use `TryFrom` and add specialised `Extend` impls, this seems to work.